php/php.editor/src/org/netbeans/modules/php/editor/verification/PSR4Hint.java

```diff
@@ -0,0 +1,259 @@
+import org.netbeans.modules.php.editor.parser.astnodes.NamespaceDeclaration;
+import org.netbeans.modules.php.editor.parser.astnodes.NamespaceName;
+import org.netbeans.modules.php.editor.parser.astnodes.TraitDeclaration;
+import org.netbeans.modules.php.editor.parser.astnodes.TypeDeclaration;
+import org.netbeans.modules.php.editor.parser.astnodes.visitors.DefaultVisitor;
+import org.openide.filesystems.FileObject;
+import org.openide.filesystems.FileUtil;
+import org.openide.util.NbBundle;
+
+/**
+ * Check class name corresponds to PSR-4 rules.
+ *
+ * @see https://www.php-fig.org/psr/psr-4/
+ */
+public abstract class PSR4Hint extends HintRule {
+
+    @Override
+    public void invoke(PHPRuleContext context, List<Hint> result) {
+        PHPParseResult phpParseResult = (PHPParseResult) context.parserResult;
+        if (phpParseResult.getProgram() != null) {
+            FileObject fileObject = phpParseResult.getSnapshot().getSource().getFileObject();
+            if (fileObject != null) {
+                if (CancelSupport.getDefault().isCancelled()) {
+                    return;
+                }
+                CheckVisitor checkVisitor = createVisitor(fileObject, context.doc);
+                phpParseResult.getProgram().accept(checkVisitor);
+                if (CancelSupport.getDefault().isCancelled()) {
+                    return;
+                }
+                result.addAll(checkVisitor.getHints());
+            }
+        }
+    }
+
+    abstract CheckVisitor createVisitor(FileObject fileObject, BaseDocument baseDocument);
+
+    public static class NamespaceDeclarationHint extends PSR4Hint {
+
+        private static final String HINT_ID = "PSR4.Hint.Namespace"; //NOI18N
+
+        @Override
+        CheckVisitor createVisitor(FileObject fileObject, BaseDocument baseDocument) {
+            return new NamespaceVisitor(this, fileObject, baseDocument);
+        }
+
+        private static final class NamespaceVisitor extends CheckVisitor {
+
+            public NamespaceVisitor(PSR4Hint psr4hint, FileObject fileObject, BaseDocument baseDocument) {
+                super(psr4hint, fileObject, baseDocument);
+            }
+
+            @Override
+            @NbBundle.Messages("PSR4WrongNamespaceNameHintText=Namespace declaration name doesn't correspond to current directory structure.")
+            public void visit(NamespaceDeclaration node) {
+                if (CancelSupport.getDefault().isCancelled()) {
+                    return;
+                }
+                NamespaceName namespaceName = node.getName();
+                if (namespaceName != null) {
+                    int endOffset = namespaceName.getEndOffset();
+                    File currentDir = getFile().getParentFile();
+                    ListIterator<Identifier> segmentsIterator = namespaceName.getSegments().listIterator(namespaceName.getSegments().size());
+                    while (segmentsIterator.hasPrevious()) {
```

Comment:
Perhaps calling `CancelSupport.getDefault().isCancelled()` in this loop could be useful?

Reply:
One iteration of the loop will be done for one part of namespace.
I expect no more than 10 iterations even if someone uses very deep structure.

Can traversing directories can take long time and getting its name take so much time it will be noticeable?

Reply:
In general, this check should be present in every loop to avoid unneeded work to be done. The check is very cheap and IO operations are slow, in general (yes, I heard about SSD :) so I think it should be there.

Reply:
Right, I added it.

Reply:
Thanks!